File: src/main/java/cl/edu/dev/solid/infrastructure/exception/ControllerAdvice.java
package cl.edu.dev.solid.infrastructure.exception;

import cl.edu.dev.solid.application.dto.ErrorDTO;
import cl.edu.dev.solid.application.exception.AlreadyExistsException;
import cl.edu.dev.solid.application.exception.NotFoundException;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

@RestControllerAdvice
public class ControllerAdvice {
    @ExceptionHandler(value = RuntimeException.class)
    public ResponseEntity<ErrorDTO> runtimeExceptionHandler(RuntimeException ex) {
        ErrorDTO errorDTO = ErrorDTO.builder()
                .code("C-500")
                .message(ex.getMessage())
                .build();
        return new ResponseEntity<>(errorDTO, HttpStatus.INTERNAL_SERVER_ERROR);
    }
    @ExceptionHandler(value = NotFoundException.class)
    public ResponseEntity<ErrorDTO> notFoundExceptionHandler(NotFoundException ex) {
        ErrorDTO errorDTO = ErrorDTO.builder()
                .code(ex.getCode())
                .message(ex.getMessage())
                .build();
        return new ResponseEntity<>(errorDTO, HttpStatus.NOT_FOUND);
    }
    @ExceptionHandler(value = AlreadyExistsException.class)
    public ResponseEntity<ErrorDTO> alreadyExistsExceptionHandler(AlreadyExistsException ex) {
        ErrorDTO errorDTO = ErrorDTO.builder()
                .code(ex.getCode())
                .message(ex.getMessage())
                .build();
        return new ResponseEntity<>(errorDTO, HttpStatus.CONFLICT);
    }
}

File: src/main/java/cl/edu/dev/solid/infrastructure/exception/ControllerAdvice.kt
package cl.edu.dev.solid.infrastructure.exception

import cl.edu.dev.solid.application.dto.ErrorDTO
import cl.edu.dev.solid.application.exception.AlreadyExistsException
import cl.edu.dev.solid.application.exception.NotFoundException
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
class ControllerAdvice {
    @ExceptionHandler(value = [RuntimeException::class])
    fun runtimeExceptionHandler(ex: RuntimeException): ResponseEntity<ErrorDTO> {
        val errorDTO: ErrorDTO = ErrorDTO(
            "C-500",
            ex.message
        )
        return ResponseEntity(errorDTO, HttpStatus.INTERNAL_SERVER_ERROR)
    }

    @ExceptionHandler(value = [NotFoundException::class])
    fun notFoundExceptionHandler(ex: NotFoundException): ResponseEntity<ErrorDTO> {
        val errorDTO: ErrorDTO = ErrorDTO(
            ex.code,
            ex.message
        )
        return ResponseEntity(errorDTO, HttpStatus.NOT_FOUND)
    }

    @ExceptionHandler(value = [AlreadyExistsException::class])
    fun alreadyExistsExceptionHandler(ex: AlreadyExistsException): ResponseEntity<ErrorDTO> {
        val errorDTO: ErrorDTO = ErrorDTO(
            ex.code,
            ex.message
        )
        return ResponseEntity(errorDTO, HttpStatus.CONFLICT)
    }
}
